File: aletheia/rag/retrieval/cumulative_summarizer.py
"""
Cumulative Summarizer Module
Implements batch incremental summarization with table/formula handling.
"""

import re
import time
from typing import List, Dict, Optional, Tuple
from aletheia.config.settings import retrieval_config
from aletheia.rag.prompts import (
    BATCH_SUMMARY_INITIAL,
    BATCH_SUMMARY_INCREMENTAL,
    TABLE_FALLBACK_PROMPT,
)


class CumulativeSummarizer:
    """
    Handles incremental batch summarization for cumulative context.

    Features:
    - Batch summarization (configurable batch size)
    - Content type detection (text/table/formula)
    - Retry mechanism with fallback
    - Table caption extraction
    - Formula LaTeX preservation
    """

    def __init__(self, generator):
        """
        Initialize summarizer.

        Args:
            generator: LLMGenerator instance for summarization
        """
        self.generator = generator
        self.batch_size = retrieval_config.batch_size
        self.max_retries = retrieval_config.max_retries
        self.timeout = retrieval_config.timeout_seconds

    def detect_content_type(self, chunk: Dict) -> str:
        """
        Detect chunk content type.

        Args:
            chunk: Paragraph dictionary with 'item_type' and 'text'

        Returns:
            "text", "table", or "formula"
        """
        # Primary: Check database field
        if "item_type" in chunk:
            item_type = chunk.get("item_type", "paragraph").lower()
            if item_type == "table":
                return "table"
            elif item_type in ["math", "formula", "equation"]:
                return "formula"

        # Fallback: Heuristic detection
        text = chunk.get("text", "")

        # Table detection
        if self._is_table(text):
            return "table"

        # Formula detection
        if self._is_formula(text):
            return "formula"

        return "text"

    def _is_table(self, text: str) -> bool:
        """Check if text looks like a table."""
        # Check for markdown table syntax
        if "|" in text and text.count("|") > 2:
            return True

        # Check for "Table" prefix
        if text.strip().startswith("Table") or text.strip().startswith("TABLE"):
            return True

        return False

    def _is_formula(self, text: str) -> bool:
        """Check if text contains mathematical formulas."""
        # LaTeX markers
        latex_markers = [r"\(", r"\[", r"$$", r"\begin{equation}"]
        if any(marker in text for marker in latex_markers):
            return True

        # Math keywords
        math_keywords = ["equation", "=", "∫", "∑", "∂", "≈", "≤", "≥"]
        # Must have multiple indicators
        indicator_count = sum(1 for keyword in math_keywords if keyword in text)
        if indicator_count >= 2:
            return True

        return False

    def extract_table_caption(self, table_chunk: Dict) -> str:
        """
        Extract table caption/title.

        Args:
            table_chunk: Table paragraph dictionary

        Returns:
            Caption string or empty string
        """
        text = table_chunk.get("text", "")

        # Pattern 1: "Table X.Y: Caption"
        pattern1 = r"Table\s+\d+\.?\d*\s*:\s*([^\n|]+)"
        match = re.search(pattern1, text, re.IGNORECASE)
        if match:
            return match.group(0).strip()

        # Pattern 2: First line before table
        lines = text.split("\n")
        for line in lines:
            if line.strip() and "|" not in line:
                return line.strip()

        return ""

    def extract_table_structure(self, table_chunk: Dict) -> Dict:
        """
        Parse table into structured format.

        Args:
            table_chunk: Table paragraph dictionary

        Returns:
            {
                'caption': str,
                'header': List[str],
                'rows': List[List[str]],
                'raw_text': str
            }
        """
        text = table_chunk.get("text", "")
        lines = text.split("\n")

        caption = self.extract_table_caption(table_chunk)
        header = []
        rows = []

        # Parse markdown table
        for line in lines:
            if "|" in line:
                cells = [cell.strip() for cell in line.split("|") if cell.strip()]

                # Skip separator lines (---)
                if all(set(cell.strip()) <= {"-", " ", ":"} for cell in cells):
                    continue

                if not header:
                    header = cells
                else:
                    rows.append(cells)

        return {"caption": caption, "header": header, "rows": rows, "raw_text": text}

    def apply_table_strategy(self, table_chunk: Dict) -> Dict:
        """
        Decoupled Indexing: Preserve table without summarization.

        Args:
            table_chunk: Table paragraph dictionary

        Returns:
            Enriched table structure
        """
        structure = self.extract_table_structure(table_chunk)

        return {
            "type": "table",
            "caption": structure["caption"],
            "header": structure["header"],
            "rows": structure["rows"],
            "content": structure["raw_text"],
            "position": table_chunk.get("paragraph_id", ""),
            "page": table_chunk.get("page_num", 0),
        }

    def apply_formula_strategy(
        self, formula_chunk: Dict, all_chunks: List[Dict], position: int
    ) -> Dict:
        """
        Context Enrichment: Preserve formula with surrounding context.

        Args:
            formula_chunk: Formula paragraph dictionary
            all_chunks: All paragraphs in sequence
            position: Index of formula_chunk in all_chunks

        Returns:
            Enriched formula structure
        """
        # Get surrounding context
        context_before = ""
        context_after = ""

        if position > 0:
            prev_chunk = all_chunks[position - 1]
            context_before = prev_chunk.get("text", "")

        if position < len(all_chunks) - 1:
            next_chunk = all_chunks[position + 1]
            context_after = next_chunk.get("text", "")

        return {
            "type": "formula",
            "context_before": context_before,
            "formula_latex": formula_chunk.get("text", ""),
            "context_after": context_after,
            "position": formula_chunk.get("paragraph_id", ""),
            "page": formula_chunk.get("page_num", 0),
        }

    def _call_llm_with_retry(self, prompt: str) -> Tuple[bool, str]:
        """
        Call LLM with retry mechanism.

        Args:
            prompt: Prompt to send to LLM

        Returns:
            (success: bool, result: str)
        """
        for attempt in range(self.max_retries):
            try:
                # Only Kimi provider is supported
                resp = self.generator.client.chat.completions.create(
                    model=self.generator.model,
                    messages=[{"role": "user", "content": prompt}],
                    temperature=0.0,
                    timeout=self.timeout,
                )
                summary = resp.choices[0].message.content
                return True, summary

            except Exception as e:
                print(
                    f"⚠️ LLM call failed (attempt {attempt + 1}/{self.max_retries}): {e}"
                )
                if attempt < self.max_retries - 1:
                    time.sleep(2**attempt)  # Exponential backoff
                continue

        return False, ""

    def _create_fallback_summary(
        self, chunks: List[Dict], chunk_type: str = "text"
    ) -> str:
        """
        Create fallback summary when LLM fails.

        Args:
            chunks: List of chunk dictionaries
            chunk_type: "text" or "table"

        Returns:
            Fallback summary string
        """
        if chunk_type == "table":
            # Use table structure: Caption + Header + First Row
            summaries = []
            for chunk in chunks:
                structure = self.extract_table_structure(chunk)
                caption = (
                    structure["caption"]
                    or f"Table at page {chunk.get('page_num', '?')}"
                )
                header = ", ".join(structure["header"][:5])  # First 5 columns
                first_row = (
                    ", ".join(structure["rows"][0][:5]) if structure["rows"] else ""
                )

                summaries.append(
                    f"{caption} - Columns: {header} - Example: {first_row}"
                )

            return " | ".join(summaries)
        else:
            # Text fallback: First 200 chars + Last 200 chars
            if len(chunks) == 0:
                return "[No content]"

            first_text = chunks[0].get("text", "")[:200]
            last_text = chunks[-1].get("text", "")[:200] if len(chunks) > 1 else ""

            return f"{first_text}... ... {last_text}"

    def summarize_batch(self, text_chunks: List[Dict]) -> str:
        """
        Perform batch incremental summarization.

        Workflow:
        1. Batch 1 (chunks 1-N): Initial summarization
        2. Batch 2 (summary_1 + chunks N+1-M): Incremental
        3. Continue until all chunks processed

        Args:
            text_chunks: List of text paragraph dictionaries (no tables/formulas)

        Returns:
            Final cumulative summary
        """
        if not text_chunks:
            return ""

        if len(text_chunks) == 1:
            # Single chunk, no need to summarize
            return text_chunks[0].get("text", "")

        running_summary = ""
        processed = 0

        while processed < len(text_chunks):
            # Get next batch
            batch_end = min(processed + self.batch_size, len(text_chunks))
            batch = text_chunks[processed:batch_end]

            if processed == 0:
                # Initial batch
                batch_text = "\n\n".join(
                    [
                        f"[Chunk {i + 1}]:\n{chunk.get('text', '')}"
                        for i, chunk in enumerate(batch)
                    ]
                )
                prompt = BATCH_SUMMARY_INITIAL.format(
                    batch_text=batch_text, num_chunks=len(batch)
                )
            else:
                # Incremental batch
                new_chunks_text = "\n\n".join(
                    [
                        f"[Chunk {processed + i + 1}]:\n{chunk.get('text', '')}"
                        for i, chunk in enumerate(batch)
                    ]
                )
                prompt = BATCH_SUMMARY_INCREMENTAL.format(
                    previous_summary=running_summary,
                    new_chunks=new_chunks_text,
                    prev_end=processed,
                    new_end=batch_end,
                )

            # Call LLM with retry
            success, summary = self._call_llm_with_retry(prompt)

            if success:
                running_summary = summary
            else:
                # Fallback
                print(f"⚠️ Fallback summary used for chunks {processed + 1}-{batch_end}")
                fallback = self._create_fallback_summary(batch, chunk_type="text")
                if running_summary:
                    running_summary = f"{running_summary}\n\n[Chunks {processed + 1}-{batch_end}]: {fallback}"
                else:
                    running_summary = fallback

            processed = batch_end

        return running_summary

    def assemble_context(
        self,
        summary: str,
        tables: List[Dict],
        formulas: List[Dict],
        current_chunk: Dict,
    ) -> str:
        """
        Assemble final enriched context.

        Args:
            summary: Text summary from batch summarization
            tables: List of table structures
            formulas: List of formula structures
            current_chunk: The target chunk (N)

        Returns:
            Formatted context string
        """
        parts = []

        # Text summary
        if summary:
            parts.append(f"=== CONTEXT SUMMARY ===\n{summary}\n")

        # Tables (preserved)
        if tables:
            parts.append("=== TABLES (Preserved) ===")
            for table in tables:
                caption = table["caption"] or f"Table at page {table['page']}"
                parts.append(f"\n[{caption}]")
                parts.append(table["content"])
            parts.append("")

        # Formulas (enriched)
        if formulas:
            parts.append("=== FORMULAS (with Context) ===")
            for formula in formulas:
                if formula["context_before"]:
                    parts.append(f"Context: {formula['context_before'][:200]}...")
                parts.append(f"\nFORMULA:\n{formula['formula_latex']}\n")
                if formula["context_after"]:
                    parts.append(f"Explanation: {formula['context_after'][:200]}...")
            parts.append("")

        # Current chunk
        parts.append(
            f"=== CURRENT CHUNK (Page {current_chunk.get('page_num', '?')}) ==="
        )
        parts.append(current_chunk.get("text", ""))

        return "\n".join(parts)

File: src/rag/retrieval/cumulative-summarizer.ts
import { getRetrievalConfig } from "../../config/settings.js";
import { BATCH_SUMMARY_INITIAL, BATCH_SUMMARY_INCREMENTAL } from "../prompts.js";

export class CumulativeSummarizer {
  private readonly generator: { client: { chat: { completions: { create: (...args: unknown[]) => Promise<unknown> } } }; model: string };
  private readonly batchSize: number;
  private readonly maxRetries: number;
  private readonly timeout: number;

  constructor(generator: { client: { chat: { completions: { create: (...args: unknown[]) => Promise<unknown> } } }; model: string }) {
    this.generator = generator;
    const config = getRetrievalConfig();
    this.batchSize = config.batchSize;
    this.maxRetries = config.maxRetries;
    this.timeout = config.timeoutSeconds;
  }

  detectContentType(chunk: Record<string, unknown>): "text" | "table" | "formula" {
    const itemType = (chunk.item_type as string ?? "paragraph").toLowerCase();
    if (itemType === "table") return "table";
    if (["math", "formula", "equation"].includes(itemType)) return "formula";

    const text = (chunk.text as string) ?? "";
    if (this.isTable(text)) return "table";
    if (this.isFormula(text)) return "formula";
    return "text";
  }

  private isTable(text: string): boolean {
    if (text.includes("|") && (text.match(/\|/g)?.length ?? 0) > 2) return true;
    if (/^Table\b/i.test(text.trim())) return true;
    return false;
  }

  private isFormula(text: string): boolean {
    const latexMarkers = ["\\(", "\\[", "$$", "\\begin{equation}"];
    if (latexMarkers.some((m) => text.includes(m))) return true;
    const mathKeywords = ["equation", "=", "∫", "∑", "∂", "≈", "≤", "≥"];
    const count = mathKeywords.filter((k) => text.includes(k)).length;
    return count >= 2;
  }

  applyTableStrategy(tableChunk: Record<string, unknown>): Record<string, unknown> {
    const text = (tableChunk.text as string) ?? "";
    const caption = this.extractTableCaption(text);
    return {
      type: "table",
      caption,
      content: text,
      position: tableChunk.paragraph_id ?? "",
      page: tableChunk.page_num ?? 0,
    };
  }

  applyFormulaStrategy(
    formulaChunk: Record<string, unknown>,
    allChunks: readonly Record<string, unknown>[],
    position: number,
  ): Record<string, unknown> {
    const contextBefore = position > 0 ? ((allChunks[position - 1].text as string) ?? "") : "";
    const contextAfter = position < allChunks.length - 1 ? ((allChunks[position + 1].text as string) ?? "") : "";
    return {
      type: "formula",
      context_before: contextBefore,
      formula_latex: (formulaChunk.text as string) ?? "",
      context_after: contextAfter,
      position: formulaChunk.paragraph_id ?? "",
      page: formulaChunk.page_num ?? 0,
    };
  }

  private extractTableCaption(text: string): string {
    const match = /Table\s+\d+\.?\d*\s*:\s*([^\n|]+)/i.exec(text);
    if (match) return match[0].trim();
    const lines = text.split("\n");
    for (const line of lines) {
      if (line.trim() && !line.includes("|")) return line.trim();
    }
    return "";
  }

  private async callLlmWithRetry(prompt: string): Promise<[boolean, string]> {
    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        const resp = await this.generator.client.chat.completions.create({
          model: this.generator.model,
          messages: [{ role: "user", content: prompt }],
          temperature: 0.0,
        } as Record<string, unknown>) as { choices: { message: { content: string } }[] };
        return [true, resp.choices[0].message.content];
      } catch (e) {
        console.error(`[summarizer] LLM call failed (attempt ${attempt + 1}/${this.maxRetries}):`, e);
        if (attempt < this.maxRetries - 1) {
          await new Promise((r) => setTimeout(r, 2 ** attempt * 1000));
        }
      }
    }
    return [false, ""];
  }

  createFallbackSummary(chunks: readonly Record<string, unknown>[], chunkType = "text"): string {
    if (chunks.length === 0) return "[No content]";

    if (chunkType === "table") {
      return chunks.map((c) => {
        const text = (c.text as string) ?? "";
        const caption = this.extractTableCaption(text) || `Table at page ${c.page_num ?? "?"}`;
        return caption;
      }).join(" | ");
    }

    const firstText = ((chunks[0].text as string) ?? "").slice(0, 200);
    const lastText = chunks.length > 1 ? ((chunks[chunks.length - 1].text as string) ?? "").slice(0, 200) : "";
    return `${firstText}... ... ${lastText}`;
  }

  async summarizeBatch(textChunks: readonly Record<string, unknown>[]): Promise<string> {
    if (textChunks.length === 0) return "";
    if (textChunks.length === 1) return (textChunks[0].text as string) ?? "";

    let runningSummary = "";
    let processed = 0;

    while (processed < textChunks.length) {
      const batchEnd = Math.min(processed + this.batchSize, textChunks.length);
      const batch = textChunks.slice(processed, batchEnd);

      let prompt: string;
      if (processed === 0) {
        const batchText = batch
          .map((chunk, i) => `[Chunk ${i + 1}]:\n${(chunk.text as string) ?? ""}`)
          .join("\n\n");
        prompt = BATCH_SUMMARY_INITIAL
          .replace("{batch_text}", batchText)
          .replace("{num_chunks}", String(batch.length));
      } else {
        const newChunksText = batch
          .map((chunk, i) => `[Chunk ${processed + i + 1}]:\n${(chunk.text as string) ?? ""}`)
          .join("\n\n");
        prompt = BATCH_SUMMARY_INCREMENTAL
          .replace("{previous_summary}", runningSummary)
          .replace("{new_chunks}", newChunksText)
          .replace("{prev_end}", String(processed))
          .replace("{new_end}", String(batchEnd));
      }

      const [success, summary] = await this.callLlmWithRetry(prompt);

      if (success) {
        runningSummary = summary;
      } else {
        const fallback = this.createFallbackSummary(batch, "text");
        runningSummary = runningSummary
          ? `${runningSummary}\n\n[Chunks ${processed + 1}-${batchEnd}]: ${fallback}`
          : fallback;
      }

      processed = batchEnd;
    }

    return runningSummary;
  }

  assembleContext(
    summary: string,
    tables: readonly Record<string, unknown>[],
    formulas: readonly Record<string, unknown>[],
    currentChunk: Record<string, unknown>,
  ): string {
    const parts: string[] = [];

    if (summary) {
      parts.push(`=== CONTEXT SUMMARY ===\n${summary}\n`);
    }

    if (tables.length > 0) {
      parts.push("=== TABLES (Preserved) ===");
      for (const table of tables) {
        const caption = (table.caption as string) || `Table at page ${table.page}`;
        parts.push(`\n[${caption}]`);
        parts.push(table.content as string);
      }
      parts.push("");
    }

    if (formulas.length > 0) {
      parts.push("=== FORMULAS (with Context) ===");
      for (const formula of formulas) {
        if (formula.context_before) {
          parts.push(`Context: ${(formula.context_before as string).slice(0, 200)}...`);
        }
        parts.push(`\nFORMULA:\n${formula.formula_latex}\n`);
        if (formula.context_after) {
          parts.push(`Explanation: ${(formula.context_after as string).slice(0, 200)}...`);
        }
      }
      parts.push("");
    }

    parts.push(`=== CURRENT CHUNK (Page ${currentChunk.page_num ?? "?"}) ===`);
    parts.push((currentChunk.text as string) ?? "");

    return parts.join("\n");
  }
}
